netbox/utilities/forms.py

```diff
@@ -561,6 +562,27 @@ def __init__(self, slug_source='name', *args, **kwargs):
         self.widget.attrs['slug-source'] = slug_source
 
 
+class TagFilterField(forms.MultipleChoiceField):
+    """
+    A filter field for the tags of a model. Only the tags used by a model are displayed.
+
+    :param model: The model of the filter
+    """
+    widget = StaticSelect2Multiple
+
+    def __init__(self, model, *args, **kwargs):
+        # Only instanitate the field if the model supports tags (i.e. hide if not)
+        if hasattr(model, 'tags'):
+            self.model = model
+
+            # Choices are fetched during form initialization
+            super().__init__(label='Tags', choices=self._choices, required=False, *args, **kwargs)
+
+    def _choices(self):
```

Comment:
Good call. Are you sure about using `_set_choices`? It's used as the [setter](https://docs.djangoproject.com/en/2.2/_modules/django/forms/fields/) for the `choices` property.

Reply:
I think so. The field's choices should only ever be the available tags, right? Whatever makes sense.

Reply:
I'm just confused because this is the setter code
```python
def _set_choices(self, value):
    # Setting choices also sets the choices on the widget.
    # choices can be any iterable, but we call list() on it because
    # it will be consumed more than once.
    if callable(value):
        value = CallableChoiceIterator(value)
    else:
        value = list(value)

    self._choices = self.widget.choices = value

choices = property(_get_choices, _set_choices)
```

so wouldn't the code in `_set_choices` only ever run on assignment, like `choices = something`. I was going for something like

```python
def __init__(self, model, *args, **kwargs):
    def get_choices():
        tags = self.model.tags.annotate(count=Count('extras_taggeditem_items')).order_by('name')
        return [(str(tag.slug), '{} ({})'.format(tag.name, tag.count)) for tag in tags]

    # Choices are fetched each time the form is initialized
    super().__init__(label='Tags', choices=get_choices, required=False, *args, **kwargs)
```

but now I'm just curious as to what you were referring to (It's probably going right over my head and I'll shortly end up with an "oooooh" moment).

Reply:
I think what you have now is fine, and probably cleaner.